parley/src/layout/cursor.rs

```diff
@@ -550,153 +474,95 @@ impl Selection {
-    }
-
-    /// Returns a new selection with the focus moved to a `Cursor`.
-    ///
-    /// If `extend` is `true` then the current anchor will be retained,
-    /// otherwise the new selection will be collapsed.
-    #[must_use]
-    pub fn maybe_extend(&self, focus: Cursor, extend: bool) -> Self {
-        if extend {
-            Self {
-                anchor: self.anchor,
-                focus,
+                self.focus
+            };
+            return Self {
+                anchor: new_focus,
+                focus: new_focus,
                 h_pos: None,
-            }
-        } else {
-            focus.into()
+            };
         }
+        self.maybe_extend(self.focus.previous_visual(layout), extend)
     }
 
-    /// Returns a new selection with the focus moved to the start of the
-    /// current line.
+    /// Returns a new selection with the focus moved to the next word in visual
+    /// order.
     ///
     /// If `extend` is `true` then the current anchor will be retained,
     /// otherwise the new selection will be collapsed.
     #[must_use]
-    pub fn line_start<B: Brush>(&self, layout: &Layout<B>, extend: bool) -> Self {
-        if let Some(line) = self.focus.path.line(layout) {
-            self.maybe_extend(
-                Cursor::from_index(layout, line.text_range().start, Affinity::Downstream),
-                extend,
-            )
-        } else {
-            *self
-        }
+    pub fn next_visual_word<B: Brush>(&self, layout: &Layout<B>, extend: bool) -> Self {
+        self.maybe_extend(self.focus.next_visual_word(layout), extend)
     }
 
-    /// Returns a new selection with the focus moved to the end of the
-    /// current line.
+    /// Returns a new selection with the focus moved to the previous word in
+    /// visual order.
     ///
     /// If `extend` is `true` then the current anchor will be retained,
     /// otherwise the new selection will be collapsed.
     #[must_use]
-    pub fn line_end<B: Brush>(&self, layout: &Layout<B>, extend: bool) -> Self {
-        if let Some(line) = self.focus.path.line(layout) {
-            self.maybe_extend(
-                Cursor::from_index(
-                    layout,
-                    line.text_range().end.saturating_sub(1),
-                    Affinity::Upstream,
-                ),
-                extend,
-            )
-        } else {
-            *self
-        }
+    pub fn previous_visual_word<B: Brush>(&self, layout: &Layout<B>, extend: bool) -> Self {
+        self.maybe_extend(self.focus.previous_visual_word(layout), extend)
     }
 
     /// Returns a new selection with the focus moved to the next line. The
     /// current horizontal position will be maintained.
     ///
     /// If `extend` is `true` then the current anchor will be retained,
-    /// otherwise the new selection will be collapsed.
+    /// otherwise the new selection will be collapsed.     
```

Comment:
Please remove trailing white space.

Reply:
There were a few instances of this.. I thought `cargo fmt --all` fixed this stuff.